python/pyarrow/_acero.pyx

```diff
@@ -0,0 +1,459 @@
+    #     "full outer": CJoinType_FULL_OUTER,
+    # }
+
+    def _set_options(
+        self, join_type, left_keys, right_keys, left_output=None, right_output=None,
+        output_suffix_for_left="", output_suffix_for_right="",
+    ):
+        cdef:
+            CJoinType c_join_type
+            vector[CFieldRef] c_left_keys
+            vector[CFieldRef] c_right_keys
+            vector[CFieldRef] c_left_output
+            vector[CFieldRef] c_right_output
+
+        # join type
+        # try:
+        #     c_join_type = self._join_type_map[join_type]
+        # except KeyError:
+        #     raise ValueError("Unsupported join type")
+        if join_type == "left semi":
+            c_join_type = CJoinType_LEFT_SEMI
+        elif join_type == "right semi":
+            c_join_type = CJoinType_RIGHT_SEMI
+        elif join_type == "left anti":
+            c_join_type = CJoinType_LEFT_ANTI
+        elif join_type == "right anti":
+            c_join_type = CJoinType_RIGHT_ANTI
+        elif join_type == "inner":
+            c_join_type = CJoinType_INNER
+        elif join_type == "left outer":
+            c_join_type = CJoinType_LEFT_OUTER
+        elif join_type == "right outer":
+            c_join_type = CJoinType_RIGHT_OUTER
+        elif join_type == "full outer":
+            c_join_type = CJoinType_FULL_OUTER
+        else:
+            raise ValueError("Unsupported join type")
+
+        # left/right keys
+        if isinstance(left_keys, str):
+            left_keys = [left_keys]
+        for key in left_keys:
+            c_left_keys.push_back(_ensure_field_ref(key))
+        if isinstance(right_keys, str):
+            right_keys = [right_keys]
+        for key in right_keys:
+            c_right_keys.push_back(_ensure_field_ref(key))
+
+        # left/right output fields
+        if left_output is not None and right_output is not None:
+            for colname in left_output:
+                c_left_output.push_back(_ensure_field_ref(colname))
+            for colname in right_output:
+                c_right_output.push_back(_ensure_field_ref(colname))
+
+            self.wrapped.reset(
+                new CHashJoinNodeOptions(
+                    c_join_type, c_left_keys, c_right_keys,
+                    c_left_output, c_right_output,
+                    _true,
+                    <c_string>tobytes(output_suffix_for_left),
+                    <c_string>tobytes(output_suffix_for_right)
+                )
+            )
+        else:
+            self.wrapped.reset(
+                new CHashJoinNodeOptions(
+                    c_join_type, c_left_keys, c_right_keys,
+                    _true,
+                    <c_string>tobytes(output_suffix_for_left),
+                    <c_string>tobytes(output_suffix_for_right)
+                )
+            )
+
+
+class HashJoinNodeOptions(_HashJoinNodeOptions):
+    """
+    Make a node which implements join operation using hash join strategy.
```

Comment:
This seems a bit sparse.  Maybe...

Make a node which joins two inputs together based on common keys.  Each row from the left input will be matched with rows from the right input where the keys are equal.  Null keys do not match other null keys.  If any key is null then that row will not match.  Which columns are output, and the behavior of rows that do not match, will depend on the join type.

inner: for each row in the left input there will be one output row per matching row in the right input.  If there are no matching rows in the right input then there will be no output row for that input row.  The output columns will be the left_keys, followed by the right_keys (these will be identical to the left_keys), followed by the left_output, followed by the right_output.

left outer: behaves the same as an inner join.  However, if there are no matching rows in the right input then there will be one output row that contains null for each column in right_keys and right_output.  A row in the right input, that has no matching row in the left input, will not be output.  The output columns will be identical to an inner join.

right outer: for each row in the right input there will be one output row per matching row in the left input.  If there are no matching rows in the left input then there will be one output row containing null in left_keys and left_output.  A row in the left input, that has no matching row in the right input, will not be output.  The output columns will be identical to an inner join.

full outer: behaves the same as a left outer join.  However, if there is a row in the right input, that has no matching row in the left input, then there will be one output row containing null for each column in left_keys and left_output.  The output columns will be identical to an inner join.

left semi: for each row in the left input there will be exactly one output row, if and only if, there is a matching row in the right input.  The output columns will be the left_keys followed by the left_output.

right semi: for each row in the right input there will be exactly one output row, if and only if, there is a matching row in the left input.  The output columns will be the right_keys followed by the right_output.

left anti: for each row in the left input there will be exactly one output row, if and only if, there is not a matching row in the right input.  The output will be the same as left semi.

right anti: for each row in the right input there will be exactly one output row, if and only if, there is not a matching row in the left input.  The output will be the same as right semi.

Reply:
Given that this suggestion is a larger addition, going to leave that for a separate PR.